### PR TITLE
feat: rewrite tracing

### LINKED ISSUE #216: Rewrite tracing via metadata
(Optionally?) store the sequence of rewrites applied on each circuit in TASO and other passes, so we can get a better insight and stats on which rules are the most impactful.

### PR DESCRIPTION
**Pull request by aborgna-q**
Optionally store metadata about the rewrites applied to a circuit.

If the `rewrite-tracing` feature is disabled, this has no effect on the runtime.

If the feature is enabled, circuits must opt-in to the tracing by calling `RewriteTracer::enable_rewrite_tracing`.

Closes #216

### CODE REVIEW

File: tket2/src/rewrite/trace.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+/// Metadata key for the circuit rewrite trace.
+pub const METADATA_REWRITES: &str = "TKET2.rewrites";
+
+/// Global read-only flag for enabling rewrite tracing.
+/// Enable it by setting the `rewrite-tracing` feature.
+///
+/// Note that circuits must be explicitly enabled for rewrite tracing by calling
+/// [`RewriteTracer::enable_rewrite_tracing`].
+pub const REWRITE_TRACING_ENABLED: bool = cfg!(feature = "rewrite-tracing");
```
Comment (ss2165) [inline tket2/src/rewrite/trace.rs:21]: is rust analyzer etc. good enough to ignore code guarded by this constant?

File: tket2/src/rewrite/trace.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    ///
+    /// Returns `None` if rewrite tracing is not enabled for this circuit.
+    #[inline]
+    fn rewrite_trace(&self) -> Option<Vec<RewriteTrace>> {
+        if !REWRITE_TRACING_ENABLED {
+            return None;
+        }
+        let meta = self.get_metadata(self.root(), METADATA_REWRITES)?;
+        let rewrites = meta.as_array()?;
+        Some(rewrites.iter().map_into().collect_vec())
```
Comment (ss2165) [inline tket2/src/rewrite/trace.rs:122]: could return a `&[RewriteTrace]`?

**Review by ss2165 — APPROVED** [review]

**Comment by ss2165** [comment]
There doesn't seem to be any key or namespace reservation happening in the metadata to denote a section as being for rewrites?

File: tket2/src/rewrite/trace.rs (hunk 1 above)
Comment (aborgna-q) [reply tket2/src/rewrite/trace.rs:21]: rust analyser checks the default features unless configured differently, bit this should work as any const flag.

File: tket2/src/rewrite/trace.rs (hunk 2 above)
Comment (aborgna-q) [reply tket2/src/rewrite/trace.rs:122]: Not possible with the `.map_into`, but we could return an `impl Iterator` once RPITIT` is here...

**Comment by aborgna-q** [comment]
There is no structure for the metadata namespace yet. There's a discussion thread for that: https://github.com/CQCL/hugr/discussions/268

**Pull request merged**